builtin/logical/ssh/path_issue.go

```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+func (b *backend) pathIssue(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role").(string)
+	role, err := b.getRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse(fmt.Sprintf("unknown role: %s", roleName)), nil
+	}
+
+	if role.KeyType != "ca" {
+		return logical.ErrorResponse("role key type \"%s\" not allowed to issue key pairs", role.KeyType), nil
+	}
+
+	keyType := data.Get("key_type").(string)
+	keyBits := data.Get("key_bits").(int)
+
+	keyTypeLengths, keyPresent := role.AllowedUserKeyTypesLengths[keyType]
+	if keyPresent {
+		var bitsPresent bool
+		for _, kb := range keyTypeLengths {
+			if keyBits == kb {
+				bitsPresent = true
+				break
+			}
+		}
+		if !bitsPresent {
+			// Also return list of allowed key bits?
+			return logical.ErrorResponse("key_bits not in list of allowed values for key_type provided"), nil
+		}
+	}
+
+	publicKey, privateKey, err := generateSSHKeyPair(b.Backend.GetRandomReader(), keyType, keyBits)
+	if err != nil {
+		return nil, err
+	}
+
+	if publicKey == "" || privateKey == "" {
+		return nil, fmt.Errorf("failed to generate or parse the keys")
+	}
+
+	// Sign key
+	userPublicKey, err := parsePublicSSHKey(publicKey)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("failed to parse public_key as SSH key: %s", err)), nil
+	}
+
+	err = b.validateSignedKeyRequirements(userPublicKey, role)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("public_key failed to meet the key requirements: %s", err)), nil
+	}
+
+	keyID, err := b.calculateKeyID(data, req, role, userPublicKey)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	certificateType, err := b.calculateCertificateType(data, role)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	var parsedPrincipals []string
+	if certificateType == ssh.HostCert {
+		parsedPrincipals, err = b.calculateValidPrincipals(data, req, role, "", role.AllowedDomains, validateValidPrincipalForHosts(role))
+		if err != nil {
+			return logical.ErrorResponse(err.Error()), nil
+		}
+	} else {
+		parsedPrincipals, err = b.calculateValidPrincipals(data, req, role, role.DefaultUser, role.AllowedUsers, strutil.StrListContains)
+		if err != nil {
+			return logical.ErrorResponse(err.Error()), nil
+		}
+	}
+
+	ttl, err := b.calculateTTL(data, role)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	criticalOptions, err := b.calculateCriticalOptions(data, role)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	extensions, err := b.calculateExtensions(data, req, role)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	privateKeyEntry, err := caKey(ctx, req.Storage, caPrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read CA private key: %w", err)
```

Comment:
I am working on the creation of the common logic function but found something a bit odd, why do some return statements return as `nil, err` and others `response, nil`? Example above.
I was trying to treat everything as `response, nil` but it seems to be breaking the tests.

Reply:
@Gabrielopesantos it depends on the type of error, I'd return both tbh.

I think these eventually get switched between `errutil.UserError` and `errutil.InternalError` based on something (keywords?). In the former case (line 159 in your diff above), this probably is already a user error. But lacking the private key is more of an internal/operator error than a user error. My 2c.

I _think_ errors from the handler are logged explicitly, whereas `logical.ErrorResponses` aren't logged. I think its maybe [here](https://github.com/hashicorp/vault/blob/7c3c89eec2db780e19d098a72343c426ae1b3536/sdk/logical/response_util.go#L17) ? Not entirely sure :-D 

And yes, there's really no rhyme or reason to one or the other sometimes.

Reply:
I don't think that errors are automatically logged.  And yeah, this pattern isn't used consistently at all.